_posts/2019-06-06-NeuralODEs.markdown

```diff
@@ -0,0 +1,388 @@
+     The function \(\tilde{\rho}_t\) is the density of the distribution of the random variable \(\phi_t^{-1}(y)\) where \(y\sim \mu\). The function \(\rho_t\) is the density of the distribution of the random variable \(\phi_t(x)\) where \(x\sim \rho\).
+     </p>
+     </details>
+
+  2. . What is the computational complexity of evaluating a determinant of a \(N\times N\) matrix, and why is that relevant in this context?
+     <details><summary>Solution</summary>
+     <p>
+     In general, the cost of computing the determinant of a \(N\times N\) matrix is \(O(N^3)\). To compute densities  transported by normalising flows, we need to compute the determinants of the Jacobians; therefore, an important feature of practical normalising flows, is that the Jacobian structure must allow an efficient computation of its determinant. See this week notes for more discussion on this.  
+     </p>
+     </details>
+
+<br />
+
+# 5 The adjoint method (and auto-diff)
+  **Motivation**: The adjoint method is a numerical method for efficiently computing the gradient of a function in numerical optimization problems. Understanding this method is essential to understand how to train ‘continuous depth’ nets. We also review the basics of Automatic Differentiation, which will help us understand the efficiency of the algorithm proposed in the NeuralODE paper. 
+
+  **Topics**:
+
+  1. Adjoint method
+  2. Auto-diff
+
+  **Required Reading**:
+
+  1. Section 8.7 from [Computational Science and Engineering](http://math.mit.edu/~gs/cse/) (CSE)
+  2. Sections 2,3 from [Automatic Differentiation in Machine Learning: a Survey](http://www.jmlr.org/papers/volume18/17-468/17-468.pdf)
+
+  **Notes**: Here is a [link](/assets/nodes_notes/week5.pdf) to our notes for the lesson. 
+  
+  **Optional Reading**:
+
+  1. [Prof. Steven G. Johnson's notes on adjoint method](http://math.mit.edu/~stevenj/notes.html)
+  
+  **Questions**:
+
+  1. Exercises 1,2,3 from Section 8.7 of CSE
+     <details><summary>Solution</summary>
+     <p>
+     Exercise 1:
+     This follows immediatly by noticing that the number of multiply-add operations of multiplying a \(N\times M\) matrix with a \(M\times P\) matrix is given by \(O(NMP)\). 
+     </p>
+     <p>
+     Exercise 2:
+     We simply need to apply the chain rule. Since \(\frac{\partial C}{\partial S} = 2S\) and \(\frac{dT}{dS} = \frac{\partial T}{\partial S} + \frac{\partial T}{\partial C}\frac{\partial C}{\partial S}\), we get \(\frac{d T}{d S} = 1 -2S\).  
+     </p>
+     <p>
+     Exercise 3:
+     This follows by Exercise 1 by seeing \(u^T\) and \(w^T\) as \(1\times N\) matrices and \(v\) as a \(N\times 1\) matrix. 
+     </p>
+     </details>
+
+  2. Consider the problem of optimizing a real-valued function \(g\) over the solution of the ODE \(y' = Ay\), \(y(0) = y_0\) at time \(T>0\): \(min_{y0, A} g(y(T))\). What is the solution of the adjoint equation?
+     <details><summary>Solution</summary>
+     <p>
+     See notes below.
```

Comment:
The notes aren't below and you should write this out.

Reply:
added